### DISCUSSION TITLE
Bevy and ECS docs

### DISCUSSION THREAD

**Discussion by CodyMWilson** [mattsoulanille/NovaJS#107] [Ideas]
Hey, I'm giving your docs a readthrough (great quick update btw, exactly what I needed). Would you be interested in me taking some of this documentation? I can probably write up some docs as I go and it will probably help me learn. 

> Nova ECS
> Nova ECS is the Entity Component System that NovaJS uses. It is heavily inspired by Bevy ECS.
> 
> What an ECS is
> Bevy and Wikipedia have good explainations, so this section is TODO(mattsoulanille).

I'm giving Bevy a readthrough - I'd never heard of it before now.

**Comment by mattsoulanille** [comment]
That would be awesome! Feel free to ask any questions you have.

The `nova_ecs` directory is probably the cleanest code (aside from maybe `novadatainterface`). It was written fairly recently (within the last 4 months). It's rather self-contained (and it's packable as an npm package via `bazel run //nova_ecs:ecs_pkg.pack`), so I might write jsdocs for its classes eventually. I can go into more detail on why I think an ECS is a good solution for the spaceflight part of NovaJS instead of a more normal class-based architecture in another post, but essentially what I've found is that an ECS allows for a lot of flexibility and code reuse without weird inheritance trees, mixins, or pluming data around. This isn't to say that an ECS should be used for everything. Self-contained things like the spaceport menus you see when landing probably don't need to be implemented in the ECS.

The `nova` directory has a fair amount of unused code that I've been gradually updating and removing (which I should probably just remove and use git to look at). The `.ts` files that don't use `snake_case` (as is recommended by the Google style guide) are likely no longer used, but the golden rule for what is and isn't being used is what's listed in the `BUILD.bazel` files. These unused `.ts` files exist because I had a few false starts in this refactor before settling on using an ECS. I'll eventually write up a doc explaining how I've factored everything, but here's a quick summary:
* `nova/src/nova_plugin/` has a `nova_ecs` plugin that handles everything that happens when flying around in a system (but doesn't actually draw the system).
* `nova/src/display/` has a `nova_ecs` plugin that draws everything that's in the system.
* `nova/src/spaceport/`, which doesn't exist yet, will probably contain the code for drawing the spaceport and buying outfits etc. It might be possible to copy a lot of the js version's spaceport code since it was mostly self-contained iirc.
* `nova/server.ts` Is the server's entry point. It has a single instance of `nova_ecs`'s `World`, which is the current system that's being served. Each Nova system will be handled by a different ecs `World`.
* `nova/src/browser.ts` is the browser's entry point. It has a single ecs `World` that gets populated by the server. I haven't decided if the browser should have a different world for each system or if the server should just update the browesr's world to match the system it's in.
* `nova/src/communication/` has a few non-snake-case files that I need to make snake_case that are used for communication between the browser and the server. `nova_ecs/plugins/multiplayer_plugin.ts` handles the actual logic of what gets sent when. This will have to be modified to support multiple channels (for different systems) or just replaced with Socket.io. It would be cool to have a WebRTC communication channel as well, since WebRTC is peer-to-peer and therefore lower latency than websockets.

The `novadatainterface` directory is mostly interface declarations, but there's also the implementation of `Gettable`, which is a simple class for asynchronously loading and caching resources.

The `novaparse` directory implements `novadatainterface/GameDataInterface.ts` in `novaparse/NovaParse.ts`. Parsing happens in two phases. 
* The first phase (`novaparse/src/resource_parsers`) parses the raw data out of the mac resource fork nova files and performs minimal if any transformations to the values (although the files themselves might not be perfectly consistent with this statement).
* The second phase (`novaparse/src/parsers` [which could perhaps use a rename]) transforms the parsed resource fork data into resources in `GameDataInterface`. It performs unit conversions on things like turn rate, rate of fire, etc. to make them easier to use in `nova/`. It's also where image bitmasks are applied, frames are enumerated in a PIXI-friendly manner, and convex hulls for collisions are generated.
* The existing parsers are rather verbose. New parsers might benefit from using [python-struct](https://www.npmjs.com/package/python-struct).

**Reply by CodyMWilson** [reply]
Cool! I really appreciate your support on learning, there are a lot of tools here that I've never heard of before (npm, yarn, bazel, typsecript). If I can summarize, is this correct?

- npm is a package manager for getting new dependancies
- yarn is a second package manager and project 'controller' that will organize files and call bazel (I've done some simple changes - yarn seems to handle builds pretty elegantly!)
- bazel is _just_ for controlling builds (doing the actual build)
- typescript is essentially the language that is called by bazel, which is why all files are .ts
- Docker is useful for deployment (I'm trying to pick up Docker for other projects atm, great tool)
- Bevy (not a tool) is being heavily utilized as a method of programming and convention
- Any I'm missing?

I have lots of questions :sweat_smile:. I guess my first questions
 - is the entirety of the current working source code containted in nova_ecs (and what does ecs stand for?)
 - Are there generalizations for directories that are entirely generated (by typescript for instance) rather than being true source-code? I know that listing every file could be tedious but maybe there's some easy generalizations based on filestructure.
 - You mentioned peer-to-peer, is that the plan for updating world files or will there be a server somewhere that you have in mind? If it is peer-to-peer, how do you plan on handling universal world-changing events (planet destruction for instance)? I've never worked on purely peer-to-peer architecture so I there are probably some conventions there I need to learn.

**Reply by mattsoulanille** [reply]
Thanks for the questions! They help me focus on explaining what's most confusing about the project.

> npm is a package manager for getting new dependancies

Yep! NovaJS used npm originally for getting dependencies, but now it uses yarn. They are very similar. I don't remember exactly why I switched to yarn, but it can apparently be faster in some cases. I think it has better caching, but I'm not sure. I particularly like its `yarn upgrade-interactive --latest` command for updating dependencies.

> yarn is a second package manager and project 'controller' that will organize files and call bazel (I've done some simple changes - yarn seems to handle builds pretty elegantly!)

Also yes. Yarn now fills the role of npm in the project. Yarn still pulls its packages from the same npm registry (npmjs.com) that npm pulls them from, so once packages are installed, the result is essentially the same as if you had used npm instead.

> bazel is just for controlling builds (doing the actual build)

Yes, mostly. Bazel controls builds and decides what needs to be rebuilt / retested when something changes. It also deploys the demo docker container to Google Cloud Run (this is done automatically when a commit is added to `main`), and it can publish nova_ecs to npmjs.com (although this isn't actually necessary for NovaJS to work). 

> typescript is essentially the language that is called by bazel, which is why all files are .ts

Yeah. At the moment, Bazel is mostly calling the typescript compiler to compile typescript, and the typescript refactor is pretty much entirely written in typescript. There's some potential for other languages that compile to js / web assembly to be used as well as long as they have up-to-date Bazel rules. Rust can be compiled to wasm with [this Bazel rule](https://github.com/bazelbuild/rules_rust/tree/main/wasm_bindgen) (I think. Haven't tried Rust in Bazel yet), and there might be reason to use it for particularly expensive computations. I've tested one of the [convex hull implementations](https://docs.rs/geo/0.4.5/geo/algorithm/convexhull/index.html) in Rust and when compiled to wasm, it was faster than all the other js implementations I could find on npm. It might be worth it to rewrite the PICT and rled parsers in rust since they're really slow right now (higher priority is running those parsers in a webworker, which should solve the slowness freezing the server). There's also an [emscripten rule](https://github.com/emscripten-core/emsdk/tree/master/bazel) available that can compile c / c++ to wasm.

> Docker is useful for deployment (I'm trying to pick up Docker for other projects atm, great tool)

Yep. Some people develop in docker to have a consistent environment, but I'm relying on Bazel for that and just using Docker for deployment. 

> Bevy (not a tool) is being heavily utilized as a method of programming and convention

Yes. During the typescript refactor, the design was gradually converging to be more and more like an ECS (which I didn't know about at the time). When I saw their docs, I though it was pretty much exactly the design I was looking for with the refactor. Bevy is written in rust, and for a few seconds, I considered switching the project to use rust so I could use bevy, but I decided that was likely a bad idea since it would make it harder for people to write mods to the game engine (more people know JavaScript than rust). It would also probably be harder to draw stuff on the screen, load assets, cause sounds to play, get user input, and generally do anything specific to the browser. It's possible that far in the future, most of the NovaJS engine will be transitioned to rust, but I don't see that happening any time soon.

I looked at a few js / ts ECS implementations, but didn't find one I liked enough to use. Most of them required you to extend a class in order to create a system or component, or they didn't have good support for type inference in the arguments to a system. Instead, I wrote `nova_ecs`, which reimplements a lot of Bevy's features and general ECS features. It's not optimized yet and can lag if you fire every weapon in the game at the same time once every frame, but there's a (probably) easy fix I can write that should help with performance (Cache the results to queries in [world](https://github.com/mattsoulanille/NovaJS/blob/main/nova_ecs/world.ts#L303) and update them when entities change. Queries are currently taking up about 60% of the time if firing all weapons once per frame. This would also help with GC since we can change the query results in place, I think). I was thinking I'd save this optimization for later, but it might be good to implement it now just to be sure it works.

To fire every weapon in the game once every frame, run this code in the browser console and press the spacebar:
```js
let ids = await gameData.ids;
for (let id of ids.Weapon) {
    // TODO: Have a way to get components by name, just for convenience.
    [...myShip.components.entries()][3][1].set(id, {count: 200, firing: false})
}
```

> Any I'm missing?

NovaJS used to use [protobufjs](https://www.npmjs.com/package/protobufjs) (through [Bazel rules](https://github.com/bazelbuild/rules_nodejs/tree/stable/examples/protobufjs)) for communication between clients and the server, but I switched back to JSON for ease of use and debugging. NovaJS uses [io-ts](https://gcanti.github.io/io-ts/) to verify that JSON messages are valid. While JSON is the default encoding, io-ts allows for any kind of custom encoding of a given type as long as you provide the encoder and decoder. Protobufs could still be used with io-ts if there's a specific message type that would benefit from them.

> is the entirety of the current working source code containted in nova_ecs (and what does ecs stand for?)

The entirety of NovaJS's Entity Component System (ECS) is stored in `nova_ecs/`. It's a self-contained library that doesn't really do anything on its own (although it does have a demo that can be run with `yarn bazel run //nova_ecs/demo`). The code that puts stuff on the screen and runs the server is in the `nova/` directory. The code that parses nova files is in the `novaparse/` directory. The `novadatainterface/` directory has TypeScript interfaces that define how the stuff running in `nova/` can access the data in the files parsed by `novadatainterface/`. 

> Are there generalizations for directories that are entirely generated (by typescript for instance) rather than being true source-code? I know that listing every file could be tedious but maybe there's some easy generalizations based on filestructure.

The only generated directory is `dist/` (as configured by `.bazelrc`). Bazel does not allow builds to affect the source code's file tree at all, so all build outputs end up somewhere in `dist/` whether it's convenient or not. For example, when you run `yarn bazel build //nova`, it lists the files it creates in `dist/` (although this isn't listing everything and doesn't include dependencies).
```
Target //nova:nova up-to-date:
  dist/bin/nova/nova.sh
  dist/bin/nova/nova_loader.js
  dist/bin/nova/nova_require_patch.js
```

> You mentioned peer-to-peer, is that the plan for updating world files or will there be a server somewhere that you have in mind? If it is peer-to-peer, how do you plan on handling universal world-changing events (planet destruction for instance)? I've never worked on purely peer-to-peer architecture so I there are probably some conventions there I need to learn.

It wouldn't be purely peer-to-peer. Each client would still load resources from a central server and tell the server what they are doing. However, if a webrtc (peer to peer) connection is successfully created between two peers, they can send updates to each other on that channel instead of through the server (but they still tell the server what they're doing). This can have impressively low latency if you're on the same or very close networks. It's an optimization that I plan on adding eventually, but it'll happen after the typescript version has feature parity with the js version.

**Comment by CodyMWilson** [comment]
> Yep! NovaJS used npm originally for getting dependencies, but now it uses yarn. They are very similar. I don't remember exactly why I switched to yarn, but it can apparently be faster in some cases. I think it has better caching, but I'm not sure. I particularly like its yarn upgrade-interactive --latest command for updating dependencies.

So, could npm be depreciated entirely in place of yarn? Or is there a structural reason not to do that (besides input work).

> Yep! NovaJS used npm originally for getting dependencies, but now it uses yarn. They are very similar. I don't remember exactly why I switched to yarn, but it can apparently be faster in some cases. I think it has better caching, but I'm not sure. I particularly like its yarn upgrade-interactive --latest command for updating dependencies.

Is that normal, or is there a different filetype if you had started using yarn instead of npm?

> It also deploys the demo docker container to Google Cloud Run (this is done automatically when a commit is added to main), and it can publish nova_ecs to npmjs.com (although this isn't actually necessary for NovaJS to work).

Is there currently a way to pull this docker container or is it private? I guess this is it? https://www.npmjs.com/package/nova_ecs? What is the benefit to having it push to npmjs.com - is it just another repository container?

> NovaJS uses io-ts to verify that JSON messages are valid. While JSON is the default encoding, io-ts allows for any kind of custom encoding of a given type as long as you provide the encoder and decoder. Protobufs could still be used with io-ts if there's a specific message type that would benefit from them.

Is there a reason you're foreseeing to use a custom encoding?

Great Q/A, I'm learning a lot :) I took a webdev class in college but I either didn't learn or retain about 90% of this.
